api/include/opentelemetry/baggage/baggage.h

```diff
@@ -0,0 +1,377 @@
+    // This is a workaround for the fact that memcpy doesn't accept a const destination.
+    nostd::unique_ptr<const char[]> CopyStringToPointer(nostd::string_view str)
+    {
+      char *temp = new char[str.size() + 1];
+      memcpy(temp, str.data(), str.size());
+      temp[str.size()] = '\0';
+      return nostd::unique_ptr<const char[]>(temp);
+    }
+  };
+
+  Baggage() : entries_(new Entry[kMaxKeyValuePairs]), num_entries_{0} {}
+
+  // Creates baggage object by extracting baggage info from header
+  static nostd::shared_ptr<Baggage> FromHeader(nostd::string_view header)
+  {
+    nostd::shared_ptr<Baggage> baggage{new Baggage()};
+
+    if (header.size() > kMaxSize)
+    {
+      return baggage;
+    }
+
+    size_t begin{0};
+    size_t end{0};
+    while (begin < header.size() && baggage->num_entries_ < kMaxKeyValuePairs)
+    {
+      // find list-member
+      end = header.find(kMembersSeparator, begin);
+      if (end == begin)
+      {
+        // character is ,
+        begin++;
+        continue;
+      }
+      if (end == std::string::npos)
+      {
+        // last list member. `end` points to end of it.
+        end = header.size() - 1;
+      }
+      else
+      {
+        // `end` points to end of current list member
+        end--;
+      }
+
+      auto list_member = header.substr(begin, end - begin + 1);
+
+      auto key_end_pos = list_member.find(kKeyValueSeparator);
+      if (key_end_pos == std::string::npos)
+      {
+        // Invalid list member, key end not found. Ignore this entry
+        begin = end + 2;
+        continue;
+      }
+
+      nostd::string_view metadata;
+      auto value_end_pos   = list_member.find(kValueMetadataSeparator);
+      size_t value_end_len = std::string::npos;
+      if (value_end_pos != std::string::npos)
+      {
+        value_end_len = value_end_pos - key_end_pos - 1;
+
+        // metadata is trimmed and kept as it is
+        metadata = TrimString(list_member.substr(value_end_pos + 1));
+      }
+
+      int err    = 0;
+      auto key   = Decode(TrimString(list_member.substr(0, key_end_pos)), err);
+      auto value = Decode(TrimString(list_member.substr(key_end_pos + 1, value_end_len)), err);
+      if (IsValidKey(key) && IsValidValue(value) && err == 0)
+      {
+        Entry entry(key, value, metadata);
+        if (end - begin + 1 < kMaxKeyValueSize)
+        {
+          (baggage->entries_.get())[baggage->num_entries_] = entry;
+          baggage->num_entries_++;
+        }
+      }
+
+      begin = end + 2;
+    }
+
+    return baggage;
+  }
+
+  // Creates string from baggage object.
+  std::string ToHeader()
+  {
+    std::string header_s;
+    for (size_t count = 0; count < num_entries_; count++)
+    {
+      if (count != 0)
+      {
+        header_s.append(",");
+      }
+      auto entry = (entries_.get())[count];
+      header_s.append(Encode(entry.GetKey()));
+      header_s.push_back(kKeyValueSeparator);
+      header_s.append(Encode(entry.GetValue()));
```

Comment:
I am not sure what style is preferred on OpenSource, but these 3 lines can be combined into 1 using "+" operator. You can use to_string for char input. Just a look-n-feel comment.